### DISCUSSION TITLE
Persian farsi tts

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#297] [Archived]
<USER>>>> i3130002<br/> <DATE> [November 13, 2020, 11:14am]<br/> <br/> Wanted to run [Mozilla TTS](https://github.com/mozilla/TTS/) on Persian<br/> text and was unable to do so. Doing this requires me to read almost<br/> every thing about it and get go deep in there. Wanted to post something<br/> in here so that if someone can help linking a google colab notebook<br/> others can also benefit from. *slash* <br/> Please write where can I start to be able to make Mozilla TTS work on<br/> Persian. (I believe there exists enough data on Mozilla common-voice and<br/> there should be some audio books out there too.)<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/persian-farsi-tts]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [November 13, 2020, 11:18am]<br/> <br/> This reads like you are asking others to do the work for you.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> i3130002<br/> <DATE> [November 13, 2020, 11:40am]<br/> <br/> You are right and I don't want people to do the work for me. All I'm<br/> asking is to give me hints and knowledge they know. If someone knows<br/> much he/she is able to guide me/and others how to approach this problem<br/> and if there exists any materials/codes.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> othiele<br/> <DATE> [November 13, 2020, 12:22pm]<br/> <br/> is right, you will have to<br/> put a lot of time into it as this is not some ready to use software. If<br/> you want to do the work, [this<br/> repo](https://github.com/repodiac/tit-for-tat/tree/master/thorsten-TTS/thorsten-v02-DDC)<br/> has everything<br/> you need to train a German model.<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [November 13, 2020, 10:14pm]<br/> <br/> , I may be able to help<br/> <br/> I work on [Rhasspy](https://github.com/rhasspy/rhasspy), a free & open<br/> source voice assistant that works offline. We're always looking to add<br/> new languages, both for speech recognition and text to speech.<br/> <br/> I have a [fork of MozillaTTS](https://github.com/rhasspy/larynx) that<br/> I've trained 6 voices with so far. If you'd like to collaborate, let me<br/> know.<br/> <br/> We may be able to use existing speech data if there's enough and it's<br/> good quality. Another option (if you have a good microphone) is to have<br/> you or another native speaker record a set of phrases that are<br/> phonetically 'rich'. I've built [a<br/> tool](https://github.com/rhasspy/gruut) to help find these phrases,<br/> though I'd need to add Persian/Farsi sounds.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> i3130002<br/> <DATE> [November 14, 2020, 8:54am]<br/> <br/> That's awesome, I emailed you so that we can start cooperating.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> khalilrhouma<br/> <DATE> [November 16, 2020, 10:42am]<br/> <br/> > If you want to do the work, [this<br/> > repo](https://github.com/repodiac/tit-for-tat/tree/master/thorsten-TTS/thorsten-v02-DDC)<br/> has everything<br/> > you need to train a German model.<br/> <br/> How could this repo helps to<br/> see the modification that sould be done to train TTS on another<br/> language?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> othiele<br/> <DATE> [November 16, 2020, 10:23am]<br/> <br/> Usually one learns by copying what others are doing and then you adapt<br/> it to your own needs. So I suggested you copy how we did it and then you<br/> can change stuff for Persian. If you already did that, what exactly is<br/> <br/> said, please don't ask us do it all for you. Btw,<br/> [ *slash* synesthesiam](<br/> offered his excellent repo as well. Same there, study and then ask<br/> detailed questions.<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [November 19, 2020, 1:06am]<br/> <br/> . If you do come<br/> across people who aren't able to do it themselves for under-served<br/> languages, please send them my way. I'm willing to do a lot of the work<br/> as long as I have a native speaker to consult with<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> othiele<br/> <DATE> [November 19, 2020, 9:38am]<br/> <br/> [ *slash* synesthesiam](<br/> Great to hear that and thanks for all the work you put into new models.<br/> Will send people your way<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [November 20, 2020, 12:34pm]<br/> <br/> Are your models open? Would you mind sharing to link on<br/> <https://github.com/mozilla/TTS/wiki/Released-Models><br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> hkalbasi<br/> <DATE> [December 16, 2020, 8:03am]<br/> <br/> I'm interested in Farsi TTS. What is the status of your work,<br/> <br/> [ *slash* synesthesiam](<br/> Do you need help?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> i3130002<br/> <DATE> [December 16, 2020, 8:40am]<br/> <br/> I did nothing and was unable to communicate much with him. I had to put<br/> off this project as of some personal problems, Though, I imagine using<br/> his github you should be able to start working on Farsi/Persian and if I<br/> could do something just email me (on gmail). *slash* <br/> Wish you luck<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [December 16, 2020, 3:11pm]<br/> <br/> Sorry I haven't been very responsive,<br/> . I should have some more<br/> time now during the holidays.<br/> <br/> I've made some progress, but I need help now<br/> (see below)<br/> <br/> I was able to find some Farsi speech data: I contacted the author of the<br/> [MirasVoice corpus](https://github.com/miras-tech/MirasVoice) and got<br/> the full set. Unfortunately, it doesn't have enough data from a single<br/> speaker. I might be able to use it in the future for Farsi speech to<br/> text in [Rhasspy](https://github.com/rhasspy/rhasspy), but it'll need a<br/> lot of pre-processing.<br/> <br/> So I will need to collect recordings from a volunteer. But first, I need<br/> to develop a set of sentences that have good phoneme coverage. I've<br/> already added Farsi phonemes to my [gruut-ipa<br/> library](https://github.com/rhasspy/gruut-ipa), and I've located a large<br/> corpus of sentences in [OSCAR](https://oscar-corpus.com/). But here's<br/> where I'm stuck: **numbers**.<br/> <br/> I use the [num2words](https://github.com/savoirfairelinux/num2words)<br/> library to convert digits into words (1 - *slash* > one), and it doesn't support<br/> Farsi yet. Would either of you<br/> or<br/> ) be able to help me add<br/> support?<br/> <br/> Once I can convert numbers to words, I can filter the OSCAR Farsi<br/> sentences and find a small set of sentences (usually *slash* < 2000) that will<br/> provide good phoneme pair examples. After that, we'll need to find a<br/> volunteer with a good microphone and a lot of patience. With this<br/> dataset, I'll be happy to train models for both MozillaTTS and my<br/> [Larynx fork](https://github.com/rhasspy/larynx).<br/> <br/> EDIT: Forgot to add one more step: filtering sentences. I usually start<br/> with a set of 2000-5000 phonetically rich sentences, and then ask<br/> volunteers to help filter out ones that don't make sense, are offensive<br/> somehow, or are something that a real native speaker would never say.<br/> This can be done by multiple people in parallel at least, but it's an<br/> important step<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> hkalbasi<br/> <DATE> [December 16, 2020, 6:59pm]<br/> <br/> I just made a [pull<br/> request](https://github.com/savoirfairelinux/num2words) for adding Farsi<br/> in num2words library.<br/> <br/> Why we don't use common voice sentences? There are near 7000 sentences<br/> in Farsi, which is reviewed and does not have numbers.<br/> <br/> My friend is volunteer for recording the dataset. But I think we don't<br/> have good microphone. Is IPhone's microphone considered good? Can we<br/> solve this by software? Please message me for more details in an instant<br/> messaging app so I can send you samples of my friend voice. (Matrix:<br/> `hkalbasi:mozilla.org` , telegram: `hkalbasi`)<br/> <br/> And another concern: In Farsi, characters like e, a, o are optional. For<br/> example کِتاب which means book is ketaab and کَتاب which is an invalid<br/> word is kataab (notice the small character moves from down to up, I hope<br/> your browser show that, but I can also send image) and everyone write<br/> کتاب without any a or e. This maybe is possible to handle by a<br/> dictionary. But the problem becomes more difficult in genitive case,<br/> which is connected in Farsi by e. For example my book in farsi is کِتابِ<br/> مَن = ketaabe man. But every one write it کتاب من and espeak read it<br/> ketaab man which is wrong. How we can handle that so our machine can<br/> read simple texts without ـَ ـِ ـُ explicity declared?<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [December 17, 2020, 12:48am]<br/> <br/> (my reply got lost for some reason, so I'm re-typing it)<br/> <br/> , I've pulled your changes<br/> into [my num2words fork](https://github.com/rhasspy/num2words). Thank<br/> you for such a quick response!<br/> <br/> > Why we don't use common voice sentences?<br/> <br/> This is a great idea. I looked at the (now) 9000 validated Farsi<br/> sentences, and ran them through my phoneme coverage analysis. It looks<br/> like we could get excellent coverage even with half ( *slash* ~4000), but the<br/> more data the better.<br/> <br/> How many sentences is your volunteer willing to read? I contacted you on<br/> Matrix; I'll have to listen to some samples from the iPhone to see if it<br/> will be good enough.<br/> <br/> > And another concern: In Farsi, characters like e, a, o are optional.<br/> <br/> I took some time this afternoon to look into this. In your example, is<br/> the 'e' in 'ketaab**e** man' pronounced? If so, none of the<br/> grapheme-to-phoneme systems I tried were able to produce the correct<br/> pronunciation.<br/> <br/> A dictionary approach can help, but I may need to do some more research<br/> if this is a common problem.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Muhammad_Mirab_Br<br/> <DATE> [February 15, 2021, 11:32am]<br/> <br/> Hi, is there any progress? I'm really looking forward to it. *slash* <br/> I'd be glad to help the progress of persian TTS, *slash* <br/> Cheers!<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [February 18, 2021, 3:18pm]<br/> <br/> 's friend<br/> had recorded about 400 phrases out of the 2400.<br/> <br/> If you're interested in recording, please let me know<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Muhammad_Mirab_Br<br/> <DATE> [February 19, 2021, 11:33am]<br/> <br/> Of course, I'd like to help<br/> BTW I'm also looking for an offline Persian speech recognition in Python<br/> and it would be awesome if someone could help<br/> <br/> [<br/> [[synesthesiam<br/> <DATE> [February 19, 2021, 2:52pm]<br/> <br/> I have about 400 hours of Persian speech data, which would be a good<br/> start for training a Kaldi speech recognition model.<br/> <br/> Unfortunately, some of the audio is not aligned with the corresponding<br/> text. I have an audio book, for example, whose chapters are PDFs. If you<br/> would be willing to help get this data split into (sentence, text)<br/> pairs, I will train the Kaldi model and add it to<br/> [Rhasspy](https://github.com/rhasspy/rhasspy).<br/> <br/> [**[next page<br/> →](https://discourse.mozilla.org/t/persian-farsi-tts/70498?page=2)**<br/> <br/> [Archived Post]<br/>

**Comment by ArmanMirzaei** [comment]
this is important, jumburi aslami might not care but neglecting language support worldwide gets our civilization destroyed if not addressed.
